src/Microsoft.AspNetCore.Razor.Language/CodeGeneration/DefaultDocumentWriter.cs

```diff
@@ -176,7 +176,14 @@ public override void VisitMethodDeclaration(MethodDeclarationIRNode node)
 
             public override void VisitFieldDeclaration(FieldDeclarationIRNode node)
             {
-                Context.Writer.WriteField(node.AccessModifier, node.Modifiers, node.Type, node.Name);
+                if (node.Annotations.ContainsKey(CommonAnnotations.InitializeTagHelperVariables))
+                {
+                    Context.TagHelperWriter.WriteDeclareTagHelperFields(Context);
+                }
+                else
+                {
+                    Context.Writer.WriteField(node.AccessModifier, node.Modifiers, node.Type, node.Name);
+                }
```

Comment:
I agree this is the ugliest part of this PR. As mentioned here https://github.com/aspnet/Razor/pull/1434/files#r121839784 those field declarations are special and they need some node to track their location in the IR tree and we can't split them into other basic nodes because all the information is not available at IR phase.